client/citydlg.cpp

```diff
@@ -1422,81 +1451,100 @@ void city_dialog::update_buy_button()
+    auto pix = get_citizen_sprite(tileset, categories[i], i, pcity);
+    painter->begin(citizen_pixmap);
+    painter->drawPixmap(painting_area, *pix, sprite_rectangle);
+    painter->end();
+  }
+}
+
 /**
    Redraws citizens for city_label (citizens_label)
  */
 void city_dialog::update_citizens()
 {
   enum citizen_category categories[MAX_CITY_SIZE];
-  int i, j, width, height;
-  QPainter p;
   int num_citizens =
       get_city_citizen_types(pcity, FEELING_FINAL, categories);
-  int w = tileset_small_sprite_width(tileset);
-  int h = tileset_small_sprite_height(tileset);
 
-  i = 1 + (num_citizens * 5 / 200);
-  w = w / i;
-  QRect source_rect(0, 0, w, h);
-  QRect dest_rect(0, 0, w, h);
-  width = w * num_citizens;
-  height = h;
+  int num_rows = num_citizens / CITIZENS_PER_ROW;
+
+  // extra incomplete row for leftover citizens
+  if (num_citizens % CITIZENS_PER_ROW > 0) {
+    num_rows += 1;
+  }
+
+  int sprite_width = tileset_small_sprite_width(tileset);
+  int sprite_height = tileset_small_sprite_height(tileset);
+
+  int canvas_width = sprite_width * MIN(CITIZENS_PER_ROW, num_citizens);
+  int canvas_height = sprite_height * num_rows;
 
   if (citizen_pixmap) {
     citizen_pixmap->detach();
     delete citizen_pixmap;
   }
 
-  citizen_pixmap = new QPixmap(width, height);
+  citizen_pixmap = new QPixmap(canvas_width, canvas_height);
+  QPainter painter;
 
-  for (j = 0, i = 0; i < num_citizens; i++, j++) {
-    dest_rect.moveTo(i * w, 0);
-    auto pix = get_citizen_sprite(tileset, categories[j], j, pcity);
-    p.begin(citizen_pixmap);
-    p.drawPixmap(dest_rect, *pix, source_rect);
-    p.end();
-  }
+  this->fill_citizens_pixmap(citizen_pixmap, &painter, categories,
+                             num_citizens);
 
   ui.citizens_label->set_city(pcity);
   ui.citizens_label->setPixmap(*citizen_pixmap);
+  ui.citizens_label->updateGeometry();
 
   lab_table[FEELING_FINAL]->setPixmap(*citizen_pixmap);
+  lab_table[FEELING_FINAL]->updateGeometry();
   lab_table[FEELING_FINAL]->setToolTip(text_happiness_wonders(pcity));
 
-  for (int k = 0; k < FEELING_LAST - 1; k++) {
-    lab_table[k]->set_city(pcity);
+  for (int i = 0; i < FEELING_LAST - 1; i++) {
+    lab_table[i]->set_city(pcity);
     num_citizens = get_city_citizen_types(
-        pcity, static_cast<citizen_feeling>(k), categories);
-
-    for (j = 0, i = 0; i < num_citizens; i++, j++) {
-      dest_rect.moveTo(i * w, 0);
-      auto pix = get_citizen_sprite(tileset, categories[j], j, pcity);
-      p.begin(citizen_pixmap);
-      p.drawPixmap(dest_rect, *pix, source_rect);
-      p.end();
-    }
+        pcity, static_cast<citizen_feeling>(i), categories);
+    this->fill_citizens_pixmap(citizen_pixmap, &painter, categories,
+                               num_citizens);
 
-    lab_table[k]->setPixmap(*citizen_pixmap);
+    lab_table[i]->setPixmap(*citizen_pixmap);
+    lab_table[i]->updateGeometry();
 
-    switch (k) {
```

Comment:
I'd have used the opportunity to move this `switch` out of the loop. It shouldn't be inside...

Reply:
Yeah, I didn't really focus on this part during the rewrite. Turns out it's just easier, shorter, and more readable to remove the `switch` and call the functions one by one.